### PR TITLE
Create deleteCoreV1CollectionNamespacedEvent test -  +1 endpoint coverage

### LINKED ISSUE #90914: Write deleteCoreV1CollectionNamespacedEvent test+promote - +1 endpoint coverage
# Progress `[7/7]`
* [x] Original approval issue: #90619
* [x]  APISnoop org-flow: [deleteCoreV1CollectionNamespacedEventTest.org](https://github.com/cncf/apisnoop/blob/master/tickets/k8s/deleteCoreV1CollectionNamespacedEventTest.org)
* [x]  Test approval issue: [kubernetes/kubernetes#90914](https://github.com/kubernetes/kubernetes/issues/90914)
* [X]  Test pr: [PR](https://github.com/kubernetes/kubernetes/pull/91230)
* [x]  Two weeks soak start date: Merge 19 Jun 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=delete.a.collection.of.events)
* [x]  Two weeks soak end date: 6 July 2020
* [x]  Test promotion pr: [#92813](https://github.com/kubernetes/kubernetes/pull/92813)

This issue is created to allow edit by @Riaankl

/wip
/hold

# Identifying an untested feature Using APISnoop
According to this APIsnoop query, there are still an endpoint which is untested.

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  description,
  kind
  FROM untested_stable_core_endpoints
  -- FROM untested_stable_endpoints
  where path not like '%volume%'
  -- and kind like ''
  and operation_id ilike 'delete%CollectionNamespacedEvent'
 ORDER BY kind,operation_id desc
 LIMIT 25
       ;
```

```
             operation_id              |        description         | kind  
---------------------------------------+----------------------------+-------
 deleteCoreV1CollectionNamespacedEvent | delete collection of Event | Event
(1 row)
```

# API Reference and feature documentation
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [Kubernetes API: v1.18 /deleteCoreV1CollectionNamespacedEvent](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#delete-collection-event-v1-core)
* [client-go - Event: DeleteCollection](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/event.go)

# The mock test
## Test outline
1. Create a set of 3 Events with a static label within a Namespace.
2. Confirm that all 3 Events with the label in the Namespaces are created.
3. Delete the set of Namespaced Events with a label via DeleteCollection.
4. Confirm that all Events with the label have been deleted.

## Test the functionality in Go
```go
package main

import (
  // "encoding/json"
  "fmt"
  "strconv"
  // "time"
  "flag"
  "os"
  v1 "k8s.io/api/core/v1"
  // "k8s.io/client-go/dynamic"
  // "k8s.io/apimachinery/pkg/runtime/schema"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  // "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
    fmt.Println(err)
    return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)

  // TEST BEGINS HERE

  // create a set of event test names
  eventTestName := [3]string{"test-event-1", "test-event-2", "test-event-3"}

  // create a set of test events with a label in the default namespace
  for i := range eventTestName {
    eventMessage :=  "This is " + eventTestName[i]
    _, err = ClientSet.CoreV1().Events("default").Create(&v1.Event{
      ObjectMeta: metav1.ObjectMeta{
        Name: eventTestName[i],
        Labels: map[string]string{"type": "Testing"},
      },
      Message: eventMessage,
      Reason: "Test",
      Type: "Normal",
      Count: 1,
    })
    if err != nil {
      fmt.Println("[error]", err)
      return
    }
    fmt.Println("[status] created", eventTestName[i])
    // time.Sleep(3 * time.Second)
  }

  // get a list of Events with a label in the default namespace
  eventsList, err := ClientSet.CoreV1().Events("default").List(metav1.ListOptions{
      LabelSelector: "type=Testing"})

  if err != nil {
    fmt.Println("[error]", err)
    return
  }

  // check that there are 3 events found
  fmt.Println("[status] Found", strconv.Itoa(len(eventsList.Items)), "events")
  if len(eventsList.Items) != 3 {
    fmt.Println("[error] Incorrect number of Events found")
    return
  }

  // delete Collection Events for the label in the default namespace
  _  = ClientSet.CoreV1().Events("default").DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{
    LabelSelector: "type=Testing"})
  fmt.Println("[status] DeleteCollection processed")

  // get a list of Events default namespace with the label
  eventsList, err = ClientSet.CoreV1().Events("default").List(metav1.ListOptions{
      LabelSelector: "type=Testing"})
  if err != nil {
    fmt.Println("[error]", err)
    return
  }

  // check that there are zero events found
  fmt.Println("[status] Found", strconv.Itoa(len(eventsList.Items)), "events")
  if len(eventsList.Items) != 0 {
    fmt.Println("[error] Still found an Event(s)")
    return
  }

  // TEST ENDS HERE

  fmt.Println("[status] complete")

}
```

```go
[status] created test-event-1
[status] created test-event-2
[status] created test-event-3
[status] Found 3 events
[status] DeleteCollection processed
[status] Found 0 events
[status] complete
```

# Verifying increase in coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

List endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%';
```

```
     useragent     |             operation_id              | hit_by_ete | hit_by_new_test 
-------------------+---------------------------------------+------------+-----------------
 live-test-writing | createCoreV1NamespacedEvent           | f          |               3
 live-test-writing | deleteCoreV1CollectionNamespacedEvent | f          |               1
 live-test-writing | listCoreV1NamespacedEvent             | f          |               2
(3 rows)
```

Display endpoint coverage change:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------+-----------------+--------------+--------------+------------------
 test_coverage |             445 |          188 |          191 |                3
(1 row)
```

Two other endpoints tested here have tests in progress to conformance already.

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 1 points****

This test is also created with the goal of conformance promotion.

### PR DESCRIPTION
**Pull request by riaankleinhans**
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
This PR adds a test to test the following untested endpoints:
deleteCoreV1CollectionNamespacedEvent 

Which issue(s) this PR fixes:
Fixes #90914

Special notes for your reviewer:
Adds +1 endpoint test coverage (good for conformance)

Does this PR introduce a user-facing change?:
NONE

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
NONE

/sig testing

### CODE REVIEW

**Comment by hh** [comment]
Needs to use a poll to check for deletion OR use the watch tooling.

File: test/e2e/framework/events/events.go [hunk 1]
```diff
@@ -33,6 +33,12 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 )
 
+const (
+ eventRetryPeriod  = 1 * time.Second
+ eventRetryTimeout = 1 * time.Minute
```
Comment (oomichi) [inline test/e2e/framework/events/events.go:38]: Indents seems odd, then golint and gfmt test jobs are failed as https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/91230/pull-kubernetes-verify/1265293441412108293

File: test/e2e/framework/events/events.go [hunk 2]
```diff
@@ -146,3 +203,21 @@ func eventOccurred(c clientset.Interface, namespace, eventSelector, msg string)
 		return false, nil
 	}
 }
+
+func deleteEventCollection(f *framework.Framework, label string) func() (bool, error) {
+ return func() (bool, error) {
+  var err error
```
Comment (oomichi) [inline test/e2e/framework/events/events.go:209]: ditto: odd indents

File: test/e2e/framework/events/events.go [hunk 3]
```diff
@@ -123,6 +129,57 @@ var _ = ginkgo.Describe("[sig-api-machinery] Events", func() {
 		}
 		framework.ExpectEqual(foundCreatedEvent, false, "should not have found test event after deletion")
 	})
+
+	ginkgo.It("should delete a collection of events", func() {
```
Comment (oomichi) [inline test/e2e/framework/events/events.go:133]: Please implement e2e test outside of `test/e2e/framework`.
`test/e2e/framework` is a place to keep shared common functions for e2e tests.

File: test/e2e/framework/events/events.go (hunk 1 above)
Comment (heyste) [reply test/e2e/framework/events/events.go]: Saw that earlier and fixed npw.

File: test/e2e/framework/events/events.go (hunk 2 above)
Comment (heyste) [reply test/e2e/framework/events/events.go]: Updated as well.

**Comment by Riaankl** [comment]
@spiffxp
@johnbelamaric
Can we get a/lgtm?

File: test/e2e/framework/events/events.go (hunk 3 above)
Comment (spiffxp) [reply test/e2e/framework/events/events.go:132]: whoops, looks like there's a test already in here as well

File: test/e2e/framework/events/events.go [hunk 4]
```diff
@@ -123,6 +128,57 @@ var _ = ginkgo.Describe("[sig-api-machinery] Events", func() {
...
+			LabelSelector: "testevent-set=true",
+		})
+		framework.ExpectNoError(err, "failed to get a list of events")
+
+		framework.ExpectEqual(len(eventList.Items), len(eventTestNames), "looking for expected number of pod templates events")
+
+		ginkgo.By("delete collection of events")
+		// confirm that delete collection does remove all events
+
+		err = wait.PollImmediate(eventRetryPeriod, eventRetryTimeout, deleteEventCollection(f, "testevent-set=true"))
```
Comment (spiffxp) [inline test/e2e/framework/events/events.go:169]: why are we deleting in a poll loop?

**Review by spiffxp — CHANGES_REQUESTED** [review]
@oomichi is right, we completely missed this earlier... please move this (and the test above) to `test/e2e/apimachinery/events.go`

File: test/e2e/framework/events/events.go (hunk 3 above)
Comment (heyste) [reply test/e2e/framework/events/events.go]: Both tests have been moved to `test/e2e/apimachinery/events.go`

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

File: test/conformance/testdata/conformance.yaml [hunk 5]
```diff
@@ -811,7 +811,7 @@
     new message, the check MUST have the update message. The event is deleted and
     MUST NOT show up when listing all events.
   release: v1.19
-  file: test/e2e/framework/events/events.go
+  file: test/e2e/apimachinery/events.go
```
Comment (spiffxp) [inline test/conformance/testdata/conformance.yaml:814]: @johnbelamaric this is why we need your /approve for a PR that's just adding a new non-conformance test, @oomichi caught that we shouldn't have added a conformance test to the test/e2e/framework directory (ref: https://github.com/kubernetes/kubernetes/pull/91230#discussion_r430713176) so it's been moved to the correct file

**Comment by oomichi** [comment]
Thanks for updating. Looks good for me now.

/lgtm

File: test/conformance/testdata/conformance.yaml (hunk 5 above)
Comment (johnbelamaric) [reply test/conformance/testdata/conformance.yaml:814]: oops. ok. has it been running from there though? I mean, do we have soak time for this?

Comment (johnbelamaric) [reply test/conformance/testdata/conformance.yaml:814]: i.e., it is added as a conformance test below. so I want to make sure it was actually running at some point as a non-conformance test. i can check the board

**Comment by johnbelamaric** [comment]
Looks like it's OK, I checked the testgrid.

/approve

**Pull request merged**